tests/test_filterLabelFixups.py

```diff
@@ -0,0 +1,118 @@
+            self.butler.get("flat.filterLabel", self.flatMinimalDataId)
+
+    def testFixingReadingOldFile(self):
+        """If we read an old flat with a complete data ID, we fix the
+        FilterLabel.
+        """
+        flatFullDataId = self.butler.registry.expandDataId(self.flatMinimalDataId)
+        flat = self.butler.get("flat", flatFullDataId)
+        self.assertEqual(flat.getFilterLabel().bandLabel, flatFullDataId["band"])
+        self.assertEqual(flat.getFilterLabel().physicalLabel, flatFullDataId["physical_filter"])
+        flatFilterLabel = self.butler.get("flat.filterLabel", flatFullDataId)
+        self.assertEqual(flatFilterLabel.bandLabel, flatFullDataId["band"])
+        self.assertEqual(flatFilterLabel.physicalLabel, flatFullDataId["physical_filter"])
+
+    def testReadingNewFileWithIncompleteDataId(self):
+        """If we try to read a new calexp with an incomplete data ID, the
+        reader should recognize that it can't check the filters and just trust
+        the file.
+        """
+        calexp = self.butler.get("calexp", self.calexpMinimalDataId)
+        calexpFilterLabel = self.butler.get("calexp.filterLabel", self.calexpMinimalDataId)
+        self.assertTrue(calexp.getFilterLabel().hasPhysicalLabel())
+        self.assertTrue(calexp.getFilterLabel().hasBandLabel())
+        self.assertEqual(calexp.getFilterLabel(), calexpFilterLabel)
+
+    def testReadingNewFileWithFullDataId(self):
+        """If we try to read a new calexp with a full data ID, the reader
+        should check the filters in the file for consistency with the data ID
+        (and in this case, find them consistent).
+        """
+        calexpFullDataId = self.butler.registry.expandDataId(self.calexpMinimalDataId)
+        calexp = self.butler.get("calexp", calexpFullDataId)
+        self.assertEqual(calexp.getFilterLabel().bandLabel, calexpFullDataId["band"])
+        self.assertEqual(calexp.getFilterLabel().physicalLabel, calexpFullDataId["physical_filter"])
+        calexpFilterLabel = self.butler.get("calexp.filterLabel", calexpFullDataId)
+        self.assertEqual(calexpFilterLabel.bandLabel, calexpFullDataId["band"])
+        self.assertEqual(calexpFilterLabel.physicalLabel, calexpFullDataId["physical_filter"])
+
+    def testReadingBadNewFileWithFullDataId(self):
+        """If we try to read a new calexp with a full data ID, the reader
+        should check the filters in the file for consistency with the data ID
+        (and in this case, find them inconsistent, which should result in
+        warnings and returning what's in the data ID).
+        """
+        calexpBadDataId = DataCoordinate.standardize(
+            self.calexpMinimalDataId,
+            band="g",
+            physical_filter="HSC-G",
+            visit_system=0,
+        )
+        self.assertTrue(calexpBadDataId.hasFull())
+        with self.assertWarns(Warning):
+            calexp = self.butler.get("calexp", calexpBadDataId)
```

Comment:
I'm surprised this works at all! Shouldn't the butler fail to get the data because the dataId is incorrect?

Reply:
That's a good question.  As far as the logic in the butler itself concerned, it only needs `{instrument, exposure, detector}` for this lookup, and hence it ignores any other values in the data ID.  And I think it's important for a lookup on just those to usually succeed, because often that's all the user knows.  Making the butler _always_ expand the data ID to look up the implied `{physical_filter, band}` values on `get`, so it can perform that check itself (and guarantee to `Formatter` code that the data ID is always expanded) is worth considering, but it's something I've rejected so far because it adds moderately expensive operations (small but potentially high-latency database queries) that are unnecessary in the common case that the data ID is just fine.  If we find we need other `Formatters` to look at expanded data ID values for reasons that won't fade away with time (as hopefully this one's reasons will), it's definitely worth considering.